test/cpp/tensorexpr/test_simplify.cpp

```diff
@@ -1401,6 +1401,738 @@ void testSimplifySymbolicMinMax() {
+    VarHandle x("x", kInt);
+    VarHandle y("y", kInt);
+    VarHandle z("z", kInt);
+    ExprHandle body =
+        Max::make(Min::make(x, y, true), Min::make(z, x, true), true);
+    ExprHandle simplified = IRSimplifier::simplify(body);
+
+    IS_NODE_WITH_NAME(Min, simplified.node(), min);
+    IS_VAR_WITH_NAME(min->lhs(), "x");
+    check_max(min->rhs(), "y", "z");
+  }
+
+  {
+    // Max(Min(y, x), Min(x, z)) => Min(x, Max(y, z))
+    VarHandle x("x", kInt);
+    VarHandle y("y", kInt);
+    VarHandle z("z", kInt);
+    ExprHandle body =
+        Max::make(Min::make(y, x, true), Min::make(x, z, true), true);
+    ExprHandle simplified = IRSimplifier::simplify(body);
+
+    IS_NODE_WITH_NAME(Min, simplified.node(), min);
+    IS_VAR_WITH_NAME(min->lhs(), "x");
+    check_max(min->rhs(), "y", "z");
+  }
+
+  {
+    // Max(Min(y, x), Min(z, x)) => Min(x, Max(y, z))
+    VarHandle x("x", kInt);
+    VarHandle y("y", kInt);
+    VarHandle z("z", kInt);
+    ExprHandle body =
+        Max::make(Min::make(y, x, true), Min::make(z, x, true), true);
+    ExprHandle simplified = IRSimplifier::simplify(body);
+
+    IS_NODE_WITH_NAME(Min, simplified.node(), min);
+    IS_VAR_WITH_NAME(min->lhs(), "x");
+    check_max(min->rhs(), "y", "z");
+  }
+
+  {
+    // Max(Min(y, x), Min(z, x)) => Max(Min(y, x), Min(z, x))
+    // When all the ops in the pattern do not have the same propagate_nans,
+    // it should not be simplified.
+    VarHandle x("x", kInt);
+    VarHandle y("y", kInt);
+    VarHandle z("z", kInt);
+    ExprHandle body =
+        Max::make(Min::make(y, x, true), Min::make(z, x, false), true);
+    ExprHandle simplified = IRSimplifier::simplify(body);
+
+    IS_NODE_WITH_NAME(Max, simplified.node(), max);
+    check_min(max->lhs(), "y", "x", true);
+    check_min(max->rhs(), "z", "x", false);
+    ASSERT_TRUE(max->propagate_nans());
+  }
+
+  {
+    // Max(5, Max(x, 8)) => Max(x, 8)
+    VarHandle x("x", kInt);
+    ExprHandle body = Max::make(5, Max::make(x, 8, true), true);
+    ExprHandle simplified = IRSimplifier::simplify(body);
+
+    check_max_w_const(simplified.node(), "x", 8);
+  }
+
+  {
+    // Max(8, Max(x, 5)) => Max(x, 8)
+    VarHandle x("x", kInt);
+    ExprHandle body = Max::make(8, Max::make(x, 5, true), true);
+    ExprHandle simplified = IRSimplifier::simplify(body);
+
+    check_max_w_const(simplified.node(), "x", 8);
+  }
+
+  {
+    // Max(Max(x, 8), 5) => Max(x, 8)
+    VarHandle x("x", kInt);
+    ExprHandle body = Max::make(Max::make(x, 8, true), 5, true);
+    ExprHandle simplified = IRSimplifier::simplify(body);
+
+    check_max_w_const(simplified.node(), "x", 8);
+  }
+
+  {
+    // Max(Max(x, 5), 8) => Max(x, 8)
+    VarHandle x("x", kInt);
+    ExprHandle body = Max::make(Max::make(x, 5, true), 8, true);
+    ExprHandle simplified = IRSimplifier::simplify(body);
+
+    check_max_w_const(simplified.node(), "x", 8);
+  }
+
+  {
+    // Max(5, Max(x, Max(y, Max(z, 8)))) => Max(Max(Max(x, 8), y), z)
```

Comment:
I moved some of these into new macros and got rid of some code. But as we discussed offline, I am going to keep the rest as verbose as is now, so that test failures point to the correct test case.